internal/auth/authenitcators.go

```diff
@@ -0,0 +1,113 @@
+package auth
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net/http"
+
+	bip32 "github.com/bitcoin-sv/go-sdk/compat/bip32"
+	ec "github.com/bitcoin-sv/go-sdk/primitives/ec"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/go-resty/resty/v2"
+)
+
+type XpubAuthenticator struct {
+	hdKey *bip32.ExtendedKey
+}
+
+func (x *XpubAuthenticator) Authenticate(r *resty.Request) error {
+	xPub, err := bip32.GetExtendedPublicKey(x.hdKey)
+	if err != nil {
+		return fmt.Errorf("failed to get extended public key: %w", err)
+	}
+	r.SetHeader(models.AuthHeader, xPub)
+	return nil
+}
+
+type XprivAuthenticator struct {
+	xpubAuth *XpubAuthenticator
+	xpriv    *bip32.ExtendedKey
+}
+
+func (x *XprivAuthenticator) Authenticate(r *resty.Request) error {
+	err := x.xpubAuth.Authenticate(r)
+	if err != nil {
+		return fmt.Errorf("failed to set xpub header: %w", err)
+	}
+	body := bodyString(r)
+	header := make(http.Header)
+	err = setSignature(&header, x.xpriv, body)
+	if err != nil {
+		return fmt.Errorf("failed to sign request with xpriv: %w", err)
+	}
+	r.SetHeaderMultiValues(header)
+	return nil
+}
+
+type AccessKeyAuthenticator struct {
+	priv *ec.PrivateKey
+	pub  *ec.PublicKey
+}
+
+func (a *AccessKeyAuthenticator) Authenticate(r *resty.Request) error {
+	r.Header.Set(models.AuthAccessKey, a.pubKeyHex())
+	body := bodyString(r)
+	sign, err := createSignatureAccessKey(a.privKeyHex(), body)
+	if err != nil {
+		return fmt.Errorf("failed to sign request with access key: %w", err)
+	}
+	setSignatureHeaders(&r.Header, sign)
+	return nil
+}
+
+func (a *AccessKeyAuthenticator) privKeyHex() string {
+	return hex.EncodeToString(a.priv.Serialize())
+}
+
+func (a *AccessKeyAuthenticator) pubKeyHex() string {
+	return hex.EncodeToString(a.pub.SerializeCompressed())
+}
+
+func bodyString(r *resty.Request) string {
```

Comment:
Could you explain what's the purpose of this function? If the method is get we set it to empty string but on other cases empty string aswell. I'm a little confused

Reply:
Currently we're not handling cases with the request body (i.e HTTP POST calls) so that's why it looks as follows. However, it will be updated when I'll add another part of the endpoint implementation with POST calls.